### PR TITLE
Features dict as first class solver definition

### PR DESCRIPTION
**Pull request by randomir**

### CODE REVIEW

File: dwave/cloud/client.py [hunk 1]
```diff
@@ -221,10 +224,17 @@ def from_config(cls, config_file=None, profile=None, client=None,
...
 
-            solver (str, default=None):
-                Default :term:`solver` to use in :meth:`~dwave.cloud.client.Client.get_solver`.
-                If undefined, :meth:`~dwave.cloud.client.Client.get_solver` will return the
-                first solver available.
+            solver (dict/str, default=None):
+                Default :term:`solver` features to use in :meth:`~dwave.cloud.client.Client.get_solver`.
+
+                Defined via dictionary of solver feature constraints
+                (see :meth:`~dwave.cloud.client.Client.get_solvers`). For backward
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py]: For backward compatibility, a solver name, as a string, is also accepted and converted to ``{"name": <solver name>}``.

File: dwave/cloud/client.py [hunk 2]
```diff
@@ -221,10 +224,17 @@ def from_config(cls, config_file=None, profile=None, client=None,
...
+                compatibility, string solver name is also accepted, and it's
+                transformed to ``{"name": <solver name>}``.
+
+                If undefined, :meth:`~dwave.cloud.client.Client.get_solver` will use
+                solver definition from environment variables, configuration file, or
+                fallback to the first available online solver.
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py]: If undefined, :meth:`~dwave.cloud.client.Client.get_solver` uses a solver definition from environment variables, a configuration file, or falls back to the first available online solver.

File: dwave/cloud/client.py [hunk 3]
```diff
@@ -311,11 +321,36 @@ def __init__(self, endpoint=None, token=None, solver=None, proxy=None,
...
+            solver_def = solver
+
+        elif isinstance(solver, six.string_types):
+            # support features dict encoded as JSON in our config INI file
+            # TODO: push this decoding to the config module, once we switch to a
+            #       richer config format (JSON or YAML)
+            try:
+                solver_def = json.loads(solver)
+            except Exception:
+                # unparseable json, assume string name for solver
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py]: Do we want to filter this up to the user, that her json is unparseable, maybe in a log message at least?

File: dwave/cloud/client.py [hunk 4]
```diff
@@ -311,11 +321,36 @@ def __init__(self, endpoint=None, token=None, solver=None, proxy=None,
...
+                # we'll deprecate this eventually, but for now just convert it to
+                # features dict (equality constraint on full solver name)
+                solver_def = dict(name__eq=solver)
+
+        else:
+            raise ValueError("Expecting a string name or features dictionary for 'solver'")
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py]: I'd already reverse these now: Expecting a features dictionary or a solver name as a string

File: dwave/cloud/client.py [hunk 5]
```diff
@@ -640,7 +638,7 @@ def solvers(self, refresh=False, **filters):
                                                     # two couplings required: (0,128) and (0,4)
                 qubits__issuperset={0, 4, 215},     # qubits 0, 4 and 215 have to exist
                 supported_problem_types__issubset={'ising', 'qubo'},
-                                                    # require both Ising and QUBO support
+                                                    # require Ising, QUBO or both to be supported
```
Comment (JoelPasvolsky) [inline dwave/cloud/client.py]: Good catch

File: dwave/cloud/config.py [hunk 6]
```diff
@@ -682,9 +688,17 @@ def load_config(config_file=None, profile=None, client=None,
             API authorization token.
 
         solver (str, default=None):
-            Default solver to use in :meth:`~dwave.cloud.client.Client.get_solver`.
```
Comment (JoelPasvolsky) [inline dwave/cloud/config.py]: "solver (dict/str, default=None):"       ?

File: dwave/cloud/config.py [hunk 7]
```diff
@@ -682,9 +688,17 @@ def load_config(config_file=None, profile=None, client=None,
...
+            :meth:`~dwave.cloud.client.Client.get_solver`, encoded as JSON.
+
+            Defined via dictionary of solver feature constraints
+            (see :meth:`~dwave.cloud.client.Client.solvers`). For backward
+            compatibility, string solver name is also accepted, and it's
+            transformed to ``{"name": <solver name>}``.
+
+            If undefined, :meth:`~dwave.cloud.client.Client.get_solver` will use
+            solver definition from environment variables, configuration file, or
+            fallback to the first available online solver.
```
Comment (JoelPasvolsky) [inline dwave/cloud/config.py]: Same as above

File: dwave/cloud/utils.py [hunk 8]
```diff
@@ -291,20 +291,22 @@ def __init__(self, maxage=None):
...
+                val = fn(*args, **kwargs)
+                self.cache[key] = dict(expires=now+self.maxage, val=val)
 
             return val
 
+        # expose the cache for testing and debugging
+        wrapper._cache = self.cache
+        wrapper._maxage = self.maxage
+
         return wrapper
```
Comment (JoelPasvolsky) [inline dwave/cloud/utils.py]: I think you have a missing EOL here, that sometimes causes problems

**Comment by randomir** [comment]
Closes #265.

**Pull request merged**